src/main/java/spoon/reflect/visitor/chain/QueryComposer.java

```diff
@@ -0,0 +1,76 @@
+/**
+ * Copyright (C) 2006-2016 INRIA and contributors
+ * Spoon - http://spoon.gforge.inria.fr/
+ *
+ * This software is governed by the CeCILL-C License under French law and
+ * abiding by the rules of distribution of free software. You can use, modify
+ * and/or redistribute the software under the terms of the CeCILL-C license as
+ * circulated by CEA, CNRS and INRIA at http://www.cecill.info.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE. See the CeCILL-C License for more details.
+ *
+ * The fact that you are presently reading this means that you have had
+ * knowledge of the CeCILL-C license and that you accept its terms.
+ */
+package spoon.reflect.visitor.chain;
+
+import spoon.reflect.declaration.CtElement;
+import spoon.reflect.visitor.Filter;
+
+/**
+ * QueryComposer contains methods, which can be used to create/compose a {@link QueryStep}
+ * It is implemented 1) by {@link CtElement} to allow creation new query and its first step
+ * 2) by {@link QueryStep} to allow creation of next query step
+ */
+public interface QueryComposer {
+
+	/**
+	 * Appends queryStep to the current (last) step of the query
+	 * and returns queryStep, which was entered. It is now the last query step
+	 *
+	 * @param queryStep
+	 * @return the entered queryStep, which is now the last step of the query
+	 */
+	<R> QueryStep<R> map(QueryStep<R> queryStep);

```

Comment:
> Is it possible to have a single map function with an appropriate common interface as parameter?

I do not know how it would look like. Please describe your idea in more details.

Why three functions are problem?

I need all three functions for these cases:

1) map(QueryStep) - it let's us to combine query chains together

2) map(Function) - it let's us to make a query based on standard functions of CtModel elements. It simply gets the input element and returned value is sent to next step. See javadoc description for details about 4 possible usages.

3) map(ChainableFunction) it let's us implement clever Filters, which automatically computes the scanning scope. For example FieldReferencesFilter, which searches for references to an field depending on field visibility.

Note: we cannot merge map(Function) and map(ChainableFunction) into one. Because Function#apply has 1 input parameter, while ChainableFunction#apply has 2 parameters. And if I would make one Interface, which would contains both `apply` functions, then it would be not possible to implement it using java 8 lambda expression, because it expects only one non default method in the interface.